text/0022-gas-spot-market.md

```diff
@@ -0,0 +1,215 @@
+
+### Deploy channels
+
+We have already mentioned the concept of a dedicated gas limit for WASM-less transfers in our description of the proposed chainspec parameters. To avoid crowding out WASM-less transfers, which are expected to be critical for platform health and public perception, there should be a separate "lane" for such transfers and this parameter specifies the width of this "lane." Selection among WASM-less transfers would work the same as it does for general deploys, but with sorting by WTP alone, since the gas reservation is fixed for these. Moreover, WASM-less transfers that are not selected for inclusion would be eligible for inclusion as general deploys, with stated WTP and fixed gas reservation.
+
+### Lifecycle of a deploy
+
+#### Construction
+
+(user-side)
+
+1. User creates a deploy in any supported manner, additionally specifying willingness to pay and gas reservation.
+    
+    a.  If the deploy is a WASM-less transfer, the gas reservation is fixed.
+
+#### Submission
+
+(user-side)
+
+1. User submits the deploy to any node as usual.
+
+#### Ordering
+
+(validator-side)
+
+1. Block proposer orders available deploys for inclusion in a protoblock.
+
+    a. Sort by WTP (greatest to least).
+    
+    b. Sort by gas reservation (greatest to least).
+
+#### Balance check & allocation
+
+(validator-side, proposal)
+
+For each channel, starting with WASM-less transfers, the following process is to be followed:
+
+1. Traverse the ordered list of deploys, checking balances.
+    
+    a. Check that originating account purse holds an amount equal to willingness to pay times gas reservation, record the remainder for this originating account purse.
+    
+    b. If sufficient tokens, include the deploy in the protoblock.
+    
+    c. Upon encountering the same originating account purse subsequently, check willingness to pay times gas reservation against remainder, and update the remainder.
+    
+    d. If sufficient tokens in the remainder, include the deploy in the protoblock.
+    
+2. Continue traversing deploy-by-deploy until the gas limit is full.
+
+3. Once gas limit cannot be filled by the next deploy with an eligible balance or remainder, skip deploys until a deploy small enough is encountered to be considered, otherwise continuing as above.
+
+The remainder for each originating account is to carry over from WASM-less transfers to general deploys.
+
+#### Payment code execution
+
+(validator-side, execution)
+
+1. Execute payment code as usual, but recording both originating account purse and purse provided by the payment code.
+
+#### Session code execution
+
+(validator-side, execution)
+
+1. Execute session code as usual, but subject to the available tokens in both the originating account purse and the provided purse.
+
+#### Charges
+
+(validator-side, execution)
+
+1. Charge the provided purse first, charging willingness to pay times gas reservation to the provided purse
```

Comment:
I believe we will deduct the payment amount and the motes/gas exchange rate (gas-price) will determine the amount of compute that is actually purchased.

Reply:
I'm afraid everyone is getting confused by the described UI, which doesn't actually matter for the critical logic, it just simplifies exposition (or so I thought).

A user can specify a 

(gas price, total payment amount) [what's currently done I think] 
(gas price, total gas reserved) [what this CEP uses, to simplify explanation of the reservation logic]
(total payment amount, total gas reserved)

You can always recover the missing piece from the other two, these all supply the same information, ultimately.

Reply:
The one advantage the described UI would have is never having to divide anything, however.